proposals/extensions.md

```diff
@@ -313,8 +313,115 @@ However, beyond implementing the language semantics it must satisfy certain requ
+  For example: `set_Property` => `<Extension>set_Property`.
+- An implementation method has type parameters derived from the extension container prepended to the type parameters of the original method.  
+- An implementation method has the same accessibility as the original method.  
+- The implementation method corresponding to a static method has the same parameters and return type as the original method. 
+- The implementation method corresponding to an instance method adds a prepended parameter to the signature of the original method. 
+  This parameter's refness, type, and name are derived from the receiver parameter declared in the relevant extension container.
+- The parameters in implementation methods refer to their own type parameters, instead of those of an extension container.  
+
+For example:
+```
+static class IEnumerableExtensions
+{
+    extension<T>(IEnumerable<T> source)
+    {
+        public void Method() { ... }
+        internal static int Property { get => ...; set => ...; }
+
+        public class NestedType<V>
+        {
+            public (T, U) NestedMethod<U>(V v) { ... }
+        }
+    }
+
+    extension(IAsyncEnumerable<int> values)
+    {
+        public async Task<int> Sum() { ... }
+    }
+
+    public void Method() { ... }
+}
+```
+is emitted as
+```
+static class IEnumerableExtensions
+{
+    public static class <Extension>ee_1<T>
+    {
+        public static <Extension>$(IEnumerable<T> source) => throw null;
+        public void Method() => throw null;
+        public static int Property { get => throw null; set => throw null; }
+
+        public class NestedType<V>
+        {
+            public (T, U) NestedMethod<U>(V v) => throw null;
+        }
+    }
+
+    public static class <Extension>ee_2
+    {
+        public static <Extension>$(IAsyncEnumerable<int> values) => throw null;
+        public Task<int> SumAsync() => throw null;
+    }
+
+    // Implementation for Method
+    public static void <Extension>Method<T>(IEnumerable<T> source) { ... }
+
+    // Implementation for Property
+    internal static int <Extension>get_Property<T>(IEnumerable<T> source) { ... }
+    internal static void <Extension>set_Property<T>(IEnumerable<T> source, int value) { ... }
+
+    // Implementation for NestedType
+    public class NestedType<V>
+    {
+        public (T, U) NestedMethod<T, U>(IEnumerable<T> source, V v) { ... }
+    }
+
+    // Implementation for SumAsync
+    public static int SumAsync(IAsyncEnumerable<int> values) { ... }
+}
+```
+
+Whenever extension members are used in source, we will emit those as reference to implementation methods.  
+For example: an invocation of `enumerableOfInt.Method()` would be emitted as a static call 
+to `IEnumerableExtensions.Method<int>(enumerableOfInt)`.  
+
 ## Open issues
 
+### Metadata
+
+- How do we associate signatures and implementations? Attribute+number? Naming convention? Using the prepended parameter name?
```

Comment:
Name and signature need to be unique for instance extension members so that should be enough to match them, right? I guess that's not true for static extension members.

Reply:
I'm not sure either. We may not need any additional mechanism to associate skeletons and implementations. Given a skeleton, we can always find the corresponding implementation by repeating the transformation process. But that may be heavy-handed